src/nrnpython/nrnpy_p2h.cpp

```diff
@@ -930,32 +930,40 @@ static Object* py_alltoall_type(int size, int type) {
         } else {  // scatter
 
             // destination counts
-            rcnt = new int[1];
-            nrnmpi_int_scatter(scnt.data(), rcnt, 1, root);
-            std::vector<char> r(rcnt[0] + 1);  // rcnt[0] can be 0
+            int rcnt = -1;
+            nrnmpi_int_scatter(scnt.data(), &rcnt, 1, root);
+            std::vector<char> r(rcnt + 1);  // rcnt can be 0
 
             // exchange
             if (nrnmpi_myid == root) {
                 sdispl = mk_displ(scnt.data());
             }
-            nrnmpi_char_scatterv(s.data(), scnt.data(), sdispl, r.data(), rcnt[0], root);
+            nrnmpi_char_scatterv(s.data(), scnt.data(), sdispl, r.data(), rcnt, root);
             if (sdispl)
                 delete[] sdispl;
 
-            if (rcnt[0]) {
-                nb::object po = unpickle(r);
-                pdest = po.release().ptr();
+            if (rcnt) {
+                pdest = unpickle(r);
             } else {
-                pdest = Py_None;
-                Py_INCREF(pdest);
+                pdest = nb::none();
             }
 
-            delete[] rcnt;
             assert(rdispl == NULL);
         }
     }
-    Object* ho = nrnpy_po2ho(pdest);
-    Py_XDECREF(pdest);
+
+    Object* ho = nrnpy_po2ho(pdest.ptr());
+
+    // The problem here is when `pdest` is a HOC object. In that case `ho` has `refcount == 2` but
+    // must be returned with refcount 0. The `ho` is contained in the `pdest` (which is why the HOC
+    // refcount is 2) and will be destroyed along with the `pdest` if its reference count is 0.
+    //
+    // Therefore, we must first DECREF `pdest` which also dereferences the HOC object and then
+    // decrement `ho->refcount`, because if we do it the otherway around the `Py_DECREF` encounters
+    // a HOC object with reference count 0 and deallocates it.
```

Comment:
@nrnhines could you please read this comment and check if it makes any sense? This changed looked harmless without these lines. However, it would reliably segfault on my machine. When I started debugging this was the closest I got to an explanation.